### ISSUE TITLE
Showcase image segmentation with external methods

### ISSUE THREAD

**Issue by giovp** [theislab/squidpy#326]
### Type of the feature

- [ ] New function in `sqduipy.im`?
- [ ] New function in `squidpy.gr`?
- [ ] New function in `squidpy.pl`?
- [ ] Change an existing functionality, such as default behavior?
- [x] Other?

Tutorials to show how to use external (pre-trained) segmentation models in Squidpy

### Description
In Squidpy, we do allow users to provide any callable for segmenting an image in the ImageContainer:
- see example here: https://squidpy.readthedocs.io/en/latest/auto_examples/image/compute_segment_fluo.html
- and docs: https://squidpy.readthedocs.io/en/latest/api/squidpy.im.segment.html#squidpy.im.segment

It would be nice to showcase how squidpy can easily interface ImageContainer with external models for cell segmentation.
Proposed models:
- cellpose: https://github.com/MouseLand/cellpose
- stardist: https://github.com/stardist/stardist
- more ?

Proposed datasets:
- `sq.datasets.visium_hne_image()` and `sq.datasets.visium_hne_adata()` (for the HnE also maybe other datasets, can be downloaded with `sc.datasets.visium_sge(..., include_hires_tiff=True)` )
- `sq.datasets.visium_fluo_image_crop()` and `sq.datasets.visium_fluo_adata_crop()`
- more ?

It would be important to consider few things:
- does the preprocessing impact the quality of the result (e.g. applying smooothing or grayscaling before) ?
- are there specific model parameters that seem to be more suitable for the data at hand (e.g. the use of multiple channels, other args). 

Of course all these evaluations are purely qualitative as we don't have a ground truth, but it should be enough.

Since everything can be done in notebooks, and there shouldn't be (🤞 ) any new code addition, I would suggest to work on this in notebooks in `squidpy_reproducibility` repo, and then after few iterations there, final notebooks can be pushed to `squidpy_tutorials` and rendered in docs. None of these tutorials will be added to CI.

Also mind that #324 will might impact this work, but I think it's fine to start now at least with having a sense of the models etc.

As discussed on MM, @SabrinaRichter @AnnaChristina would you mind taking a look at this? Thank you!
Happy to discuss anything missing, pinging @hspitzer and @michalk8 for additional ideas/suggestions

**Comment by michalk8** [comment]
> Since everything can be done in notebooks, and there shouldn't be (🤞 ) any new code addition

I'd say this should depend on 2 things:
1. how easily interfaceable are the methods (just seen for `cellpose`, also related to the Z-Dim to be added in `ImageContainer`:
  `For multi-channel, multi-Z tiff's, the expected format is Z x channels x Ly x Lx.`
2. how much convenience we want to create (don't think wrapping would be a problem + having a good performing seg. model is always a plus)

> Also mind that #324 will might impact this work, but I think it's fine to start now at least with having a sense of the models etc.

From API standpoint, it should be the same (i.e. input is a `y,x,c` image, Z-Dim to be added sometime later.
The only difference is that cropping will be done through dask.

**Comment by giovp** [comment]
> I'd say this should depend on 2 things:

how easily interfaceable are the methods (just seen for cellpose, also related to the Z-Dim to be added in ImageContainer:
For multi-channel, multi-Z tiff's, the expected format is Z x channels x Ly x Lx.

ok so this for sure wouldn't change current API/standards to support it. meaning, if model expects transposed dimensions, than we can just show in the notebooks how to do that, without providing a method in ImageCOntainer

> how much convenience we want to create (don't think wrapping would be a problem + having a good performing seg. model is always a plus)

you mean wrapping it as putting it into core squidpy? This would make segmentation models+tensor libraries to be dependencies. I wouldn't do it as there are always new methods comning up and who knows maybe in 6 motnhs a vision transformer will be SOTA...
but of course happy to discuss @michalk8 @hspitzer 

> The only difference is that cropping will be done through dask.

nice yes, as discussed in #324 should be fine to let user handles chunks

**Comment by hspitzer** [comment]
I agree with Giovanni that I would not change the API to accommodate these methods -- Ideally all of these transformations of the crops could be done by applying a "preprocessing" fn before pushing it in the segmentation model. 

Depending on how well the segmentation works, we could add this code in squidpy.external. This way we avoid making TF etc core dependencies, but the users would need minimal code to use the segmentation. 
Call could e.g. be like : `sq.im.segment(img, method=sq.external.cellpose)`

But first, we should see how well we can get this to work and then worry about putting it in core/external squidpy code.

**Comment by giovp** [comment]
> Ideally all of these transformations of the crops could be done by applying a "preprocessing" fn before pushing it in the segmentation model.

agree =, although some of these methods already have their own "preprocessing" so also need to be carefull here, prob just allowing to pass crops that belong to any imgcontainer layer is already good.

> But first, we should see how well we can get this to work and then worry about putting it in core/external squidpy code.

also agree, actually from @SabrinaRichter first try it looks very simple to interface cellpose https://github.com/theislab/squidpy_reproducibility/pull/11

the external module could even be potentially overkill... but let's say after testing more

**Comment by giovp** [comment]
this was closed with `squidpy=1.1.0`
